Refactor fixclock hook so doesn't run on LXC

JedMeister opened the pull request:
Refactor fixclock hook so doesn't run on LXC (with appropriate warning message)

Also requires https://github.com/turnkeylinux/buildtasks/pull/78 be merged

Closes https://github.com/turnkeylinux/tracker/issues/1885

JedMeister committed:
Refactor fixclock hook so doesn't run on LXC (with appropriate warning message)

Christie018 reviewed (approved):
Code changes look good to me!

a3s7p commented on contrib/fixclock-hook:
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
+
+    returncode, stdout, stderr = runit(['systemd-detect-virt', '-c'])
+
+    if stdout.strip() == 'none':
```
exit code should be checked here instead, and then if it's 0, the container type can be read

> If a virtualization technology is detected, 0 is returned, a non-zero code otherwise.

https://www.freedesktop.org/software/systemd/man/latest/systemd-detect-virt.html

a3s7p commented on contrib/fixclock-hook:
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
+
+    if stdout.strip() == 'none':
+        # "proper" VM or bare metal
+        return 'no'
```
you can just return (and the caller can check against) `None` here if you use `Optional[str]` from `typing` as the return type

a3s7p replied:
https://docs.python.org/3/library/typing.html#typing.Optional

a3s7p commented on contrib/fixclock-hook:
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
 
-def main():
+def dont_fixclock() -> None:
+    retcode, container, stderr = runit(['systemd-detect-virt'])
```
this call to `systemd-detect-virt` should probably be replaced with passing the return value of `is_container()` to this function

a3s7p commented on contrib/fixclock-hook:
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
-        fixclock()
+        if is_container() == 'no':
+            fixclock()
+        else:
```
the container type return value of `is_container()` is never used

JedMeister replied:
Yeah good call man. I've refactored things a bit so now there is only one call to `systemd-detect-virt` and the container type is passed to `dont_fixclock()` for inclusion in the message (a nice bonus is that the logic in `main()` now reads better too):
```
-def dont_fixclock() -> None:
-    retcode, container, stderr = runit(['systemd-detect-virt'])
-    msg = Template(LXC_TPL).substitute(CONTAINER=container.strip())
+def dont_fixclock(container: str) -> None:
+    msg = Template(CONT_TPL).substitute(CONTAINER=container)
     echo(msg, stderr=True)
 
 
@@ -113,10 +112,11 @@ def main() -> None:
     op, state = sys.argv[1:]
 
     if op in ('restore', 'backup') and state == 'pre':
-        if is_container() == 'no':
+        container = is_container()
+        if not container:
             fixclock()
         else:
-            dont_fixclock()
+            dont_fixclock(container)
```

a3s7p commented on contrib/fixclock-hook:
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
 
-    command = ["ntpdate", "-u", NTPSERVER]
+def echo(msg: str, stderr: bool = False, fatal: bool = False) -> None:
+    """ Echo messge to terminal (exit 1 if fatal)"""
```
typo

a3s7p commented on contrib/fixclock-hook:
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
 $ERROR
+"""
+
+LXC_TPL = f"""\
```
should it be called `LXC_TPL` if any container type is detected?

JedMeister replied:
Noticed that myself. I'll change it.

a3s7p reviewed (changes requested):
otherwise lgtm

a3s7p commented on contrib/fixclock-hook:
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
+Amazon S3 and Duplicity need a UTC synchronized clock, but TKLBAM detected
+that we're running in a container ($CONTAINER).
 
+As the container host manages the local time (and an unprivileged container
```
if the container has the required privileges, it can change the host clock. i think it is still best to avoid doing that even if we can — principle of least surprise. but then maybe reword the message a bit to clarify that this hook doesn't run inside containers as a matter of principle

JedMeister replied:
Ok thanks man. I'll tweak the text a little to make it correct.

JedMeister committed:
Apply suggested shellcheck tweaks

JedMeister commented:
Thanks @a3s7p - Applied your suggested changes, plus also ran shellcheck over the bash bits.

JedMeister commented:
Argh! I didn't push my last commit (where I actually addressed your points) to this branch before merging...

So I just cherry-picked it into master and pushed: https://github.com/turnkeylinux/tklbam/commit/ed83184c15468faae25a2757074d039de73f135a

a3s7p commented:
@JedMeister  just noticed another nitpick: https://github.com/turnkeylinux/tklbam/commit/ed83184c15468faae25a2757074d039de73f135a#r136367478

JedMeister commented:
Thanks @a3s7p - as per direct responses to your comments, I have put the `strip()`s back in.

Whilst not ideal, I again just pushed straight to master: https://github.com/turnkeylinux/tklbam/commit/131a741b0f104227835ef60c4e8739e4add57727
